ambry-store/src/main/java/com/github/ambry/store/ScanResults.java

```diff
@@ -26,46 +26,124 @@
+  // is expired, it's invalidated right aways. So when calculating valid size for log segment, we have to pass two time
+  // stamps, a delete reference time, usually is now - 7 days, and a expiry time, usually is now.
+  private final NavigableMap<String, Long> logSegmentBaseBucket = new TreeMap<>();
+  private final NavigableMap<Long, NavigableMap<String, Long>> logSegmentBucketsDeltaForDeleted = new TreeMap<>();
+  private final NavigableMap<Long, NavigableMap<String, Long>> logSegmentBucketsDeltaForExpired = new TreeMap<>();
+
+  final long logSegmentForecastStartTimeMsForDeleted;
+  final long logSegmentLastBucketTimeMsForDeleted;
+  final long logSegmentForecastEndTimeMsForDeleted;
+  final long logSegmentForecastStartTimeMsForExpired;
+  final long logSegmentLastBucketTimeMsForExpired;
+  final long logSegmentForecastEndTimeMsForExpired;
   Offset scannedEndOffset = null;
 
   /**
    * Create the bucket data structures in advance based on the given scanStartTime and segmentScanTimeOffset.
    */
   ScanResults(long startTimeInMs, long logSegmentForecastOffsetMs, int bucketCount, long bucketSpanInMs) {
-    long containerBucketTimeMs = startTimeInMs;
-    long logSegmentBucketTimeMs = startTimeInMs - logSegmentForecastOffsetMs;
-    for (int i = 0; i < bucketCount; i++) {
+    // Set up container buckets
+    containerForecastStartTimeMs = startTimeInMs;
+    long containerBucketTimeMs = containerForecastStartTimeMs + bucketSpanInMs;
+    for (int i = 1; i < bucketCount; i++) {
       containerBuckets.put(containerBucketTimeMs, new HashMap<>());
-      logSegmentBuckets.put(logSegmentBucketTimeMs, new TreeMap<>(LogSegmentNameHelper.COMPARATOR));
       containerBucketTimeMs += bucketSpanInMs;
-      logSegmentBucketTimeMs += bucketSpanInMs;
     }
-    containerForecastStartTimeMs = containerBuckets.firstKey();
-    containerLastBucketTimeMs = containerBuckets.lastKey();
+    containerLastBucketTimeMs = containerBuckets.isEmpty() ? containerForecastStartTimeMs : containerBuckets.lastKey();
     containerForecastEndTimeMs = containerLastBucketTimeMs + bucketSpanInMs;
-    logSegmentForecastStartTimeMs = logSegmentBuckets.firstKey();
-    logSegmentLastBucketTimeMs = logSegmentBuckets.lastKey();
-    logSegmentForecastEndTimeMs = logSegmentLastBucketTimeMs + bucketSpanInMs;
+
+    // Set up log segment buckets
+    logSegmentForecastStartTimeMsForExpired = startTimeInMs;
+    long bucketTimeMs = logSegmentForecastStartTimeMsForExpired + bucketSpanInMs;
+    for (int i = 1; i < bucketCount; i++) {
+      logSegmentBucketsDeltaForExpired.put(bucketTimeMs, new TreeMap<>(LogSegmentNameHelper.COMPARATOR));
+      bucketTimeMs += bucketSpanInMs;
+    }
+    logSegmentLastBucketTimeMsForExpired =
+        logSegmentBucketsDeltaForExpired.isEmpty() ? logSegmentForecastStartTimeMsForExpired
+            : logSegmentBucketsDeltaForExpired.lastKey();
+    logSegmentForecastEndTimeMsForExpired = logSegmentLastBucketTimeMsForExpired + bucketSpanInMs;
```

Comment:
`containerForecastEndTimeMs`,  `logSegmentForecastEndTimeMsForExpired` and  `logSegmentForecastEndTimeMsForDeleted` all seem to be extending another `bucketSpanInMs` from last bucket time. Can you explain why? We can talk this offline if needed.